File: app/src/main/java/com/esoxjem/movieguide/details/MovieDetailsPresenter.java
package com.esoxjem.movieguide.details;

import com.esoxjem.movieguide.Movie;
import com.esoxjem.movieguide.Review;
import com.esoxjem.movieguide.Video;
import com.esoxjem.movieguide.favorites.IFavoritesInteractor;
import com.esoxjem.movieguide.util.RxUtils;

import java.util.List;

import rx.Subscriber;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * @author arun
 */
class MovieDetailsPresenter implements IMovieDetailsPresenter
{
    private IMovieDetailsView view;
    private IMovieDetailsInteractor movieDetailsInteractor;
    private IFavoritesInteractor favoritesInteractor;
    private Subscription trailersSubscription;
    private Subscription reviewSubscription;

    MovieDetailsPresenter(IMovieDetailsInteractor movieDetailsInteractor, IFavoritesInteractor favoritesInteractor)
    {
        this.movieDetailsInteractor = movieDetailsInteractor;
        this.favoritesInteractor = favoritesInteractor;
    }

    @Override
    public void setView(IMovieDetailsView view)
    {
        this.view = view;
    }

    @Override
    public void destroy()
    {
        view = null;
        RxUtils.unsubscribe(trailersSubscription, reviewSubscription);
    }

    @Override
    public void showDetails(Movie movie)
    {
        if (isViewAttached())
        {
            view.showDetails(movie);
        }
    }

    private boolean isViewAttached()
    {
        return view != null;
    }

    @Override
    public void showTrailers(Movie movie)
    {
        trailersSubscription = movieDetailsInteractor.getTrailers(movie.getId()).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<List<Video>>()
                {
                    @Override
                    public void onCompleted()
                    {
                        // do nothing
                    }

                    @Override
                    public void onError(Throwable e)
                    {
                        // do nothing
                    }

                    @Override
                    public void onNext(List<Video> videos)
                    {
                        onGetTrailersSuccess(videos);
                    }
                });
    }

    private void onGetTrailersSuccess(List<Video> videos)
    {
        if (isViewAttached())
        {
            view.showTrailers(videos);
        }
    }

    @Override
    public void showReviews(Movie movie)
    {
        reviewSubscription = movieDetailsInteractor.getReviews(movie.getId()).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Subscriber<List<Review>>()
                {
                    @Override
                    public void onCompleted()
                    {
                        // do nothing
                    }

                    @Override
                    public void onError(Throwable e)
                    {
                        // do nothing
                    }

                    @Override
                    public void onNext(List<Review> reviews)
                    {
                        onGetReviewsSuccess(reviews);
                    }
                });
    }

    private void onGetReviewsSuccess(List<Review> reviews)
    {
        if (isViewAttached())
        {
            view.showReviews(reviews);
        }
    }

    @Override
    public void showFavoriteButton(Movie movie)
    {
        boolean isFavorite = favoritesInteractor.isFavorite(movie.getId());
        if (isViewAttached())
        {
            if (isFavorite)
            {
                view.showFavorited();
            } else
            {
                view.showUnFavorited();
            }
        }
    }

    @Override
    public void onFavoriteClick(Movie movie)
    {
        if (isViewAttached())
        {
            boolean isFavorite = favoritesInteractor.isFavorite(movie.getId());
            if (isFavorite)
            {
                favoritesInteractor.unFavorite(movie.getId());
                view.showUnFavorited();
            } else
            {
                favoritesInteractor.setFavorite(movie);
                view.showFavorited();
            }
        }
    }
}

File: app/src/main/java/com/esoxjem/movieguide/movies/details/MovieDetailsPresenter.kt
package com.esoxjem.movieguide.movies.details

import com.esoxjem.movieguide.movies.entities.Movie
import com.esoxjem.movieguide.movies.entities.Review
import com.esoxjem.movieguide.movies.entities.Video
import com.esoxjem.movieguide.movies.favorites.FavoritesContract
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

/**
 * @author arunsasidharan
 */
class MovieDetailsPresenter
(private val movieDetailsInteractor: MovieDetailsContract.Interactor,
 private val favoritesInteractor: FavoritesContract.Interactor) : MovieDetailsContract.Presenter {

    private var view: MovieDetailsContract.View? = null
    private val compositeDisposable = CompositeDisposable()


    override fun setView(view: MovieDetailsContract.View) {
        this.view = view
    }

    override fun destroy() {
        view = null
        compositeDisposable.clear()
    }

    override fun showDetails(movie: Movie) {
        view?.showDetails(movie)
    }

    override fun showTrailers(movie: Movie) {
        val trailersSubscription = movieDetailsInteractor.getTrailers(movie.id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(::onGetTrailersSuccess, ::onGetTrailersFailure)
        compositeDisposable.add(trailersSubscription)
    }

    private fun onGetTrailersSuccess(videos: List<Video>) {
        view?.showTrailers(videos)
    }

    private fun onGetTrailersFailure(t: Throwable) {
        // Do nothing
    }

    override fun showReviews(movie: Movie) {
        val reviewSubscription = movieDetailsInteractor.getReviews(movie.id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(::onGetReviewsSuccess, ::onGetReviewsFailure)
        compositeDisposable.add(reviewSubscription)
    }

    private fun onGetReviewsSuccess(reviews: List<Review>) {
        view?.showReviews(reviews)
    }

    private fun onGetReviewsFailure(t: Throwable) {
        // Do nothing
    }

    override fun showFavoriteButton(movie: Movie) {
        val isFavorite = favoritesInteractor.isFavorite(movie.id)
        if (isFavorite) {
            view?.showFavorited()
        } else {
            view?.showUnfavorited()
        }
    }

    override fun onFavoriteClick(movie: Movie) {
        val isFavorite = favoritesInteractor.isFavorite(movie.id)
        if (isFavorite) {
            favoritesInteractor.unFavorite(movie.id)
            view?.showUnfavorited()
        } else {
            favoritesInteractor.setFavorite(movie)
            view?.showFavorited()
        }
    }
}
